### PR TITLE
Fix vectorization with reduction domains in fusion inputs

### LINKED ISSUE #658: Vectorization failure with reduction domains in fusion inputs
Found while cleaning up the vectorization helper (#654). See https://github.com/NVIDIA/Fuser/pull/654#discussion_r1275723443

In `NVFuserTest.FusionSegmentReduceSoftmax_CUDA`, the fusion has a standard reduction followed by a softmax. It's a test to exercise segmentation such that the first reduction comprises a first segment and the softmax does a second segment.

The issue is that we fail to vectorize the second softmax kernel even though the problem size is large enough and all inputs and outputs are contiguous. Here's the heuristic parameters:

```
===== Reduction Stats ========
total_reduction_numel: 64
total_iteration_numel: 32
inner_most_dimension_numel: 64
vectorize_factor: 1
n_tensor_inputs: 1
max_input_dtype_size: 8
max_persistent_buffer_size: 512
max_multi_reduction_factor: 1
block(8, 1, 1)
===== Reduction Parameters ========
Tag: Inner Persistent Heuristic.

Red On Fastest Dim
Persistent Kernel
Batches per block: 4

Iteration Domain: blockIdx.x /
Inner Reduction Domain: cross block - threadIdx.x / persistent batch - 4 / unroll / factor 2
Launch Parameters: BlockDim.x = -1, BlockDim.y = 1, BlockDim.z = -1, GridDim.x = -1, GridDim.y = -1, GridDim.z = -1, Smem Size = 0

====================================
```

The reason is we hit [this condition](https://github.com/NVIDIA/Fuser/pull/654#discussion_r1275723443), so at that point we give up vectorizing the fusion.

### PR DESCRIPTION
**Pull request by naoyam**
Issue #658 is because we are using the number of consumer domains to compute a position in its producer tensor. Since a producer may have reduction IDs that the consumer do not have, the computed position can result in larger than intended, making fewer number of vectorized innermost domains. This issue happens only with the reduction and normalization schedulers.

Fixes #658

### CODE REVIEW

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Review by zasdfgbnm — APPROVED** [review]
Thanks for the fix!

File: csrc/scheduler/vectorize_helper.cpp [hunk 1]
```diff
@@ -91,6 +91,14 @@ ContiguousInnerDimensionsMapper::ContiguousInnerDimensionsMapper(
   // cannot consider that a contiguous merge dimension for vectorization.
   std::vector<IterDomain*> reordered_rfactor;
   for (auto id : reference->getMaybeRFactorDomain()) {
+    // Exclude reduction IDs if the reference is a fusion input as they
+    // don't manifest at all in the fusion. This simplifies the
+    // analysis in getContigMergeOfInnerSize, which only looks at
+    // non-reduction rfactor domains. Including reduction domains here
+    // can result in incorrect ordering
+    if (reference->isFusionInput() && id->isReduction()) {
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/scheduler/vectorize_helper.cpp:99]: Makes sense. Added an assertion

**Comment by naoyam** [comment]
!build

**Pull request merged**